buildpacks/dotnet/src/errors.rs

```diff
@@ -348,16 +406,322 @@ fn log_io_error(header: &str, occurred_while: &str, io_error: &io::Error) {
+        ));
+    }
+
+    #[test]
+    fn test_sdk_layer_verify_archive_checksum_error() {
+        assert_error_snapshot(&DotnetBuildpackError::SdkLayer(
+            SdkLayerError::VerifyArchiveChecksum {
+                expected: vec![0xAA, 0xBB, 0xCC, 0xDD],
+                actual: vec![0x11, 0x22, 0x33, 0x44],
+            },
+        ));
+    }
+
+    #[test]
+    fn test_sdk_layer_open_archive_error() {
+        assert_error_snapshot(&DotnetBuildpackError::SdkLayer(SdkLayerError::OpenArchive(
+            create_io_error(),
+        )));
+    }
+
+    #[test]
+    fn test_sdk_layer_read_archive_error() {
+        assert_error_snapshot(&DotnetBuildpackError::SdkLayer(SdkLayerError::ReadArchive(
+            create_io_error(),
+        )));
+    }
+
+    #[test]
+    fn test_parse_buildpack_configuration_invalid_msbuild_verbosity_level_error() {
+        assert_error_snapshot(&DotnetBuildpackError::ParseBuildpackConfiguration(
+            DotnetBuildpackConfigurationError::InvalidMsbuildVerbosityLevel("Foo".to_string()),
+        ));
+    }
+
+    #[test]
+    fn test_parse_buildpack_configuration_unsupported_execution_environment_error() {
+        assert_error_snapshot(&DotnetBuildpackError::ParseBuildpackConfiguration(
+            DotnetBuildpackConfigurationError::ExecutionEnvironmentError(
+                ExecutionEnvironmentError::UnsupportedExecutionEnvironment("foo".to_string()),
+            ),
+        ));
+    }
+
+    #[test]
+    fn test_restore_dotnet_tools_command_system_error() {
+        assert_error_snapshot(&DotnetBuildpackError::RestoreDotnetToolsCommand(
+            fun_run::CmdError::SystemError(
+                "Failed to start process".to_string(),
+                create_io_error(),
+            ),
+        ));
+    }
+
+    #[test]
+    fn test_restore_dotnet_tools_command_non_zero_exit_not_streamed_error() {
+        assert_error_snapshot(&DotnetBuildpackError::RestoreDotnetToolsCommand(
+            create_cmd_error(1),
+        ));
+    }
+
+    #[test]
+    fn test_publish_command_system_error() {
+        assert_error_snapshot(&DotnetBuildpackError::PublishCommand(
+            fun_run::CmdError::SystemError("Cannot find executable".to_string(), create_io_error()),
+        ));
+    }
+
+    #[test]
+    fn test_publish_command_non_zero_exit_not_streamed_error() {
+        assert_error_snapshot(&DotnetBuildpackError::PublishCommand(create_cmd_error(5)));
+    }
+
+    #[test]
+    fn test_copy_runtime_files_error() {
+        assert_error_snapshot(&DotnetBuildpackError::CopyRuntimeFiles(create_io_error()));
+    }
+
+    fn assert_error_snapshot(error: &DotnetBuildpackError) {
+        assert_writer_snapshot(|writer| on_buildpack_error_with_writer(error, writer));
+    }
+
+    fn assert_writer_snapshot(function: impl FnOnce(&mut dyn Write)) {
+        let output = {
+            let mut buffer = Vec::new();
+            function(&mut buffer);
+            String::from_utf8(buffer).unwrap()
```

Comment:
Consider removing ansi escape codes from the snapshots. The escape codes mostly add noise and, for the purpose of these snapshots, the content is more important that the format.

```suggestion
            bullet_stream::ansi_escape::strip_ansi(String::from_utf8(buffer).unwrap())
```

Reply:
I did consider it (and noticed that you're stripping the ansi codes from your tests). They're an important part of the output though, reliably printed, and easy to regenerate if changed in the future.

Why is it noise in your opinion? I can see they're not as pretty when viewed on GitHub, but that seems like more of a presentation issue? I often use `cat` to view files like these locally.

It might make sense to strip ansi escape codes etc from some snapshots though -- they're pretty similar because I'm only testing errors right now. But it'll be even more relevant when testing some of the other output, warnings etc.

Reply:
Because the formatting is often inferable from the content without the noise of the color codes. For example:
- styled values have the format `{value}` + color codes
- error lines have the format `! {line}` + color codes 

I agree that it would be nice to capture all the formatting + colors but the visual noise it adds for reviewers is a real nuisance. 

I think a nice middle ground that would preserve the complete formatting and relieve the burden from reviewers of remembering that `[0;31m!` is red (error) or `[1;36m` is blue (info) would be to convert the escape codes into a format that captures the intent, not the implementation. E.g.;

```html
- <format type="info">Debug info</format>
  - foo bar baz

<format type="error">Error completing buildpack detection</format>
<format type="error"></format>
<format type="error">An unexpected I/O error occurred while determining if we must run the Heroku .NET buildpack for this application..</format>
<format type="error"></format>
<format type="error">Use the debug information above to troubleshoot and retry your build. If the</format>
<format type="error">issue persists, file an issue here:</format>
<format type="error">https://github.com/heroku/buildpacks-dotnet/issues/new</format>
```

Not sure if that's better or worse though. For now, I think stripping the ANSI escapes provides the best balance of benefits / tradeoffs.

Reply:
I think rewriting the output into a whole new format wouldn't address why I want to test the actual output from the buildpack in the first place - it'd just be another rewrite that could introduce new errors, and not let me test what I'd like. I'm probably a bit more interested in verifying those because the classic buildpack [uses the ANSI codes when reformatting the log output](https://github.com/heroku/heroku-buildpack-dotnet/blob/4d00b3ed469195daa76f2cf273f5307e61726724/bin/compile#L77-L126) from this buildpack.

I don't think I follow entirely why the snapshot/reference value needs to capture the intent beyond what is literally the intended output. The code that produces this output [expresses the intent fairly accurately](https://github.com/heroku/buildpacks-dotnet/blob/main/buildpacks/dotnet/src/errors.rs#L351-L363), and I'd much rather reviewers look there for intent.

Anyway I hear you on wanting to reduce noise for reviewers. As I mentioned earlier, I think the _accurate_ snapshots included in this PR appearing "noisy" is a presentation issue first and foremost. From my perspective, a noisy test assertion is one that will cause intermittent test failures, unnecessary test implementation churn to meet changing needs, and/or be difficult/painful to write and/or maintain. I have yet to see flakiness from the assertions implemented here, the assertions were easy to set up, and simple to regenerate if changes are made to for instance the `bullet_stream` implementation (which I'd definitely prefer to know about with test failures after I update that dependency).

With that in mind, I think I have a much better solution to address the review aspect: GitHub Actions renders output from commands like they appear to most users, including ANSI codes, coloring etc - so we can "piggyback" on that by writing the snapshot content to the output of an action. This allows reviewers to inspect the asserted output as it'll be presented to users, and lets the tests verify the actual output without compromises.

I've added [a simple `print-insta-snapshots` action](https://github.com/heroku/buildpacks-dotnet/pull/253/commits/4afb2d4cbf3c021fdc2f939257b23b43aa58a435), and you can [see the result here](https://github.com/heroku/buildpacks-dotnet/actions/runs/14425671671/job/40454158220#step:3:29):

* It searches the entire repository for `.snap` files.
* For each snapshot file it creates [an expandable log group](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/workflow-commands-for-github-actions#grouping-log-lines) and prints the contents of the file within that group.

This seems more elegant to me, and the collapsed overview of all snapshots may even improve the experience for reviewers (who could just have the result of the `print-insta-snapshots` action open in a separate window while reading through code changes). If something sticks out, the full path to the snapshot file is readily available, so PR comments can be made in a file included in the easily (we could print a link to the actual snapshot file for added convenience).

<img width="762" alt="print-insta-snapshot-github-action" src="https://github.com/user-attachments/assets/73fbfa2b-7a28-4ed3-bcad-13886f9e8490" />

Reply:
I've updated the action a bit so it now:

* Only runs on pull request events.
* Only prints output for files that were changed in the PR (compared to the base ref).
* Includes a direct link to the corresponding snapshot file in the PR diff context.

This should allow reviewers to easily inspect changes, add comments/suggestions as needed, without introducing noise or inconvenience to the process.

You can see the [latest result here](https://github.com/heroku/buildpacks-dotnet/actions/runs/14427361536/job/40457970260#step:3:51)

Reply:
> I think the accurate snapshots included in this PR appearing "noisy" is a presentation issue first and foremost.

This is essentially what I was getting at @runesoerensen, and perhaps I didn't communicate that well. There's a discrepancy between what you see as rendered terminal output and what we see rendered in the browser and, unfortunately, a reviewer is not going to be reviewing the output in a terminal window so they'll get the messier representation 😅 

Maybe a GitHub Action would help. Though, like my intermediate format idea, it does suffer from drawbacks too. Namely, that the primary review tool here is a GitHub PR. There's no way to comment on the GHA output directly so you'd have to trace what you're looking at back to the source and that indirection introduces inconvenience.

Reply:
Did you see [my last comment](https://github.com/heroku/buildpacks-dotnet/pull/253#discussion_r2041066401)? I think that addresses your points here?

Reply:
I'm not sure I understand it otherwise... If I push a PR and include a link to the rendered output (from the GitHub action, which is part of the PR, and links back to it), that seems pretty similar to including a link to "Rendered" markdown content (as is fairly common in PR)?

Considering each rendered snapshot includes a link to the specific file *in the PR view context*, the level of indirection here is really minimal and it's straightforward for any reviewer to comment. Considering this solution lets the buildpack test what needs to be test, and presents the output exactly like the user will see it (rather than a stripped or rewritten output), I think the drawbacks here are more agreeable.

Reply:
Your GHA solution is _a solution_ but I still think there's room for improvement here. Regardless, we don't need to solve this specific problem in this PR. 

If you want to include the ANSI codes in your snapshots (with or without the associated GHA) then go for it. This discussion should not block that from moving forward.

Reply:
Ok that sounds good to me -- should I maybe just remove the GitHub action then and resolve this issue? Any other changes I should make before this can be approved? :)

Reply:
👍 no other changes needed.

As for the GitHub action, your call. Keep it if you think it'll add value. Or toss it into a branch if you want to keep it as an idea for how we can reduce that difference between the accuracy of what we output vs. presenting it for review in a different medium.

Reply:
Alright cool I've removed the GitHub action from this branch, and stored them in a local branch.

I think you need to resolve this conversation (it won't matter if I do it as it's a code suggestion, which I believe required the commenter to resolve it (if the change isn't commited))

Reply:
Thanks!